waku/v2/protocol/waku_filter_v2/protocol.nim

```diff
@@ -0,0 +1,184 @@
+## Waku Filter protocol for subscribing and filtering messages
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/[options,sequtils,sets,tables],
+  chronicles,
+  chronos,
+  libp2p/peerid,
+  libp2p/protocols/protocol
+import
+  ../../node/peer_manager,
+  ../waku_message,
+  ./common,
+  ./protocol_metrics,
+  ./rpc_codec,
+  ./rpc
+
+logScope:
+  topics = "waku filter"
+
+const
+  MaxSubscriptions* = 1000 # TODO make configurable
+  MaxCriteriaPerSubscription = 1000
+
+type
+  FilterCriterion* = (PubsubTopic, ContentTopic) # a single filter criterion is fully defined by a pubsub topic and content topic
+  FilterCriteria* = HashSet[FilterCriterion] # a sequence of filter criteria
+
+  WakuFilter* = ref object of LPProtocol
+    subscriptions*: Table[PeerID, FilterCriteria] # a mapping of peer ids to a sequence of filter criteria
+    peerManager: PeerManager
+
+proc pingSubscriber(wf: WakuFilter, peerId: PeerID): FilterSubscribeResult =
+  trace "pinging subscriber", peerId=peerId
+
+  if peerId notin wf.subscriptions:
+    debug "pinging peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.NOT_FOUND,
+      cause: "peer has no subscriptions"
+    ))
+
+  ok()
+
+proc subscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
+  if pubsubTopic.isNone() or contentTopics.len() == 0:
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.BAD_REQUEST,
+      cause: "pubsubTopic and contentTopics must be specified"
+    ))
+
+  let filterCriteria = toHashSet(contentTopics.mapIt((pubsubTopic.get(), it)))
+
+  trace "subscribing peer to filter criteria", peerId=peerId, filterCriteria=filterCriteria
+
+  if peerId in wf.subscriptions:
+    var peerSubscription = wf.subscriptions.mgetOrPut(peerId, initHashSet[FilterCriterion]())
+    if peerSubscription.len() + filterCriteria.len() >= MaxCriteriaPerSubscription:
+      return err(FilterSubscribeError(
+        kind: FilterSubscribeErrorKind.SERVICE_UNAVAILABLE,
+        cause: "peer has reached maximum number of filter criteria"
+      ))
```

Comment:
> In my opinion, we shouldn't.

Agreed! I also considered this scenario and think a request should either be fully honoured or fail (partially honouring a request is a nightmare in error handling from both client and service node's perspective)

> here, despite erroring out, we are inserting in the subscriptions map

Wait, I don't see it or am missing something obvious. 😅 We return in L63 without doing the insert in L68 and L69. Good idea to return earlier though.

> e.g. MAX 20 subscriptions per query

Good point. Will also clarify this as a recommendation so that we get reasonable filter requests from the beginning.

> I would calculate the remaining "subscription slots

As mentioned above, sounds reasonable to me too to return earlier here, although I don't quite see that we're currently exceeding max criteria per peer subscription. Will address in a follow-up (the subscriptions map is likely to be split to a separate module soon in any case, as I can imagine we may want to make it a bit more complex to e.g. build a reverse index, add timestamps, a sweeper, etc.)

Reply:
> Wait, I don't see it or am missing something obvious.  We return in L63 without doing the insert in L68 and L69. Good idea to return earlier though.

My bad, I misread this part. I thought that `mgetOrPut` was modifying already the table's value. True. You bail out before setting the value back to the table.